core/src/main/scala/kafka/log/Log.scala

```diff
@@ -246,7 +246,8 @@ class Log(@volatile var dir: File,
     // The earliest leader epoch may not be flushed during a hard failure. Recover it here.
     _leaderEpochCache.clearAndFlushEarliest(logStartOffset)
 
-    loadProducerState(logEndOffset, reloadFromCleanShutdown = hasCleanShutdownFile)
+    assert(producerStateManager.isEmpty, "Producer state must be empty during log initialization")
```

Comment:
Yeah, this could use some refactoring. I am not fully certain if it is okay to pollute `Log.producerStateManager` before this point. So far, we avoid doing that - for example, `recoverSegmentAndReloadProducerState` uses a local reference to build up the producer state which is subsequently thrown away.

Reply:
I think it would be more conventional to raise `IllegalStateException`. Usually we prefer to have checks which cannot be elided.